File: android/app/src/main/kotlin/net/defined/mobile_nebula/APIClient.kt
package net.defined.mobile_nebula

class InvalidCredentialsException: Exception("Invalid credentials")

class APIClient(context: android.content.Context) {
    private val packageInfo = PackageInfo(context)
    private val client = mobileNebula.MobileNebula.newAPIClient(
        "MobileNebula/%s (Android %s)".format(
                packageInfo.getVersion(),
                packageInfo.getSystemVersion(),
        ))

    fun enroll(code: String): String {
        val res = client.enroll(code)
        return res.site
    }

    fun tryUpdate(siteName: String, hostID: String, privateKey: String, counter: Long, trustedKeys: String, nebulaCert: String, nebulaKey: String): String? {
        val res: mobileNebula.TryUpdateResult
        try {
            res = client.tryUpdate(siteName, hostID, privateKey, counter, trustedKeys, nebulaCert, nebulaKey)
        } catch (e: Exception) {
            // type information from Go is not available, use string matching instead
            if (e.message == "invalid credentials") {
                throw InvalidCredentialsException()
            }

            throw e
        }

        if (res.fetchedUpdate) {
            return res.site
        }

        return null
    }
}

File: ios/Runner/APIClient.swift
import MobileNebula

enum APIClientError: Error {
  case invalidCredentials
}

class APIClient {
  let apiClient: MobileNebulaAPIClient

  init() {
    let packageInfo = PackageInfo()
    apiClient = MobileNebulaNewAPIClient(
      "MobileNebula/\(packageInfo.getVersion()) (iOS \(packageInfo.getSystemVersion()))")!
  }

  func enroll(code: String) throws -> String {
    let res = try apiClient.enroll(code)
    return res.site
  }

  func tryUpdate(
    siteName: String, hostID: String, privateKey: String, counter: Int, trustedKeys: String
  ) throws -> String? {
    let res: MobileNebulaTryUpdateResult
    do {
      res = try apiClient.tryUpdate(
        siteName,
        hostID: hostID,
        privateKey: privateKey,
        counter: counter,
        trustedKeys: trustedKeys)
    } catch {
      // type information from Go is not available, use string matching instead
      if error.localizedDescription == "invalid credentials" {
        throw APIClientError.invalidCredentials
      }

      throw error
    }

    if res.fetchedUpdate {
      return res.site
    }

    return nil
  }
}
